### PR TITLE
Add serialization support for Client and Dataset 

### PR DESCRIPTION
**Pull request by billschereriii**
C, C++, Fortran get to_string() methods
C++ also gets ostream& operator<<() methods
Python gets __str__() methods

Sample output (Client):

```python
>>> from smartredis import Client
>>> c = Client("127.0.0.1:6381", cluster=True, logger_name="testcase"); print(c)
Client (testcase):
Clustered Redis connection:
  Redis shards at:
    127.0.0.1:6383
    127.0.0.1:6381
    127.0.0.1:6382
  Protocol: TCP
  Command parameters:
    Retry attempts: 101
    Retry interval (ms): 1000
    Attempt timeout (ms): 100
  Connection parameters:
    Retry attempts: 101
    Retry interval (ms): 1000
    Attempt timeout (ms): 100
  Threadpool: 4 threads
```

Sample output (DataSet):

```python
>>> from smartredis import Dataset
>>> import numpy as np
>>> d = Dataset("This is a test")
>>> data = np.uint8([2,4,8])
>>> d.add_tensor("u8_tensor", data)
>>> data = np.double([2.0,4.1,8.3, 5.6])
>>> d.add_tensor("double_tensor", data)
>>> d.add_meta_scalar("float2_scalar", np.float(3.1415926535))
>>> d.add_meta_scalar("float_scalar", np.double(3.1415926535))
>>> d.add_meta_string("metastring", "metavalue")
>>> print(d)
DataSet (This is a test):
Tensors:
  double_tensor:
    type: double
    dimensions: [4]
    elements: 4
  u8_tensor:
    type: 8 bit unsigned integer
    dimensions: [3]
    elements: 3
Metadata:
  metastring:
    type: string
  float2_scalar:
    type: double
  float_scalar:
    type: double
```

### CODE REVIEW

File: src/python/bindings/bind.cpp [hunk 1]
```diff
@@ -54,6 +52,7 @@ PYBIND11_MODULE(smartredisPy, m) {
         .def(py::init<std::string&>());
 
     // Python client class
+    #define CLIENT_METHOD(name)  def(#name, &PyClient::name)
```
Comment (MattToast) [inline src/python/bindings/bind.cpp:55]: Is this macro used? (Like the idea of using it)

File: src/python/bindings/bind.cpp [hunk 2]
```diff
@@ -112,9 +111,11 @@ PYBIND11_MODULE(smartredisPy, m) {
         .def("poll_list_length_gte", &PyClient::poll_list_length_gte)
         .def("poll_list_length_lte", &PyClient::poll_list_length_lte)
         .def("get_datasets_from_list", &PyClient::get_datasets_from_list)
-        .def("get_dataset_list_range", &PyClient::get_dataset_list_range);
+        .def("get_dataset_list_range", &PyClient::get_dataset_list_range)
+        .def("to_string", &PyClient::to_string);
 
     // Python Dataset class
+    #define DATASET_METHOD(name) def(#name, &PyClient::name)
```
Comment (MattToast) [inline src/python/bindings/bind.cpp:118]: Same "macro used" question here

File: tests/python/test_dataset_methods.py [hunk 3]
```diff
@@ -27,6 +27,19 @@
...
+    """
+    dataset = Dataset("test-dataset")
+    data = np.uint8([2,4,8])
+    dataset.add_tensor("u8_tensor", data)
+    data = np.double([2.0,4.1,8.3, 5.6])
+    dataset.add_tensor("double_tensor", data)
+    dataset.add_meta_scalar("float2_scalar", float(3.1415926535))
+    dataset.add_meta_scalar("float_scalar", np.double(3.1415926535))
+    dataset.add_meta_string("metastring", "metavalue")
+    print(dataset)
```
Comment (MattToast) [inline tests/python/test_dataset_methods.py:41]: Little worried about this test: technically speaking, w/o the changes in this PR, this should still pass.

Maybe just check to make sure that the string of a dataset is no equal to the repr (which is the default when casting an object to a str)

```py
dataset = Dataset(...)
...
assert repr(dataset) != str(dataset)
```

Or you could search for a sub-string that we expect to be in the "string-ified" dataset

```py
assert "Dataset(test-dataset):" in str(dataset)
```

File: tests/python/test_address.py [hunk 4]
```diff
@@ -28,6 +28,12 @@
 
 from smartredis import Client
 
+def test_serialization(use_cluster, context):
+    # get env var to set through client init
+    ssdb = os.environ["SSDB"]
+    c = Client(address=ssdb, cluster=use_cluster, logger_name=context)
+    print(c)
```
Comment (MattToast) [inline tests/python/test_address.py:35]: Similar worry to dataset test: this test should pass reguardless of changes made in this PR. Maybe do the same `repr(c) != str(c)` trick?

**Review by MattToast — CHANGES_REQUESTED** [review]
Looked through the `*.py/pybind/Py*.cpp` files in detail and have a couple of small notes, but otherwise everything looks good. Skimmed through the reset and didn't see any immediate red flags.

Overall looks great!! This should be very helpful for users attempting to debug their simulations!!

File: src/python/bindings/bind.cpp (hunk 1 above)
Comment (billschereriii) [reply src/python/bindings/bind.cpp:55]: Not yet -- it will be a separate PR to switch over to it

File: src/python/bindings/bind.cpp (hunk 2 above)
Comment (billschereriii) [reply src/python/bindings/bind.cpp:118]: Not yet -- it will be a separate PR to switch over to it

File: tests/fortran/client_test_dataset.F90 [hunk 5]
```diff
@@ -195,6 +196,11 @@ program main
   if (result .ne. SRNoError) error stop
   if (.not. all(meta_int64_recv == meta_int64_vec)) error stop 'meta_int64: FAILED'
 
+  ! Test dataset serialization
+  write (*,*) dataset%to_string()
```
Comment (ashao) [inline tests/fortran/client_test_dataset.F90:200]: Might be better to compare this to what the expected string should be just to make sure `to_string` isn't just returning garbage

File: tests/python/test_dataset_methods.py (hunk 3 above)
Comment (billschereriii) [reply tests/python/test_dataset_methods.py]: Done

File: tests/python/test_address.py (hunk 4 above)
Comment (billschereriii) [reply tests/python/test_address.py]: Done

File: src/fortran/client/client_interfaces.inc [hunk 6]
```diff
@@ -152,3 +152,11 @@ interface
     logical(kind=c_bool)          :: exists
   end function poll_key_c
 end interface
+
+interface
+  function client_to_string_c(client) bind(c, name="client_to_string")
+    use iso_c_binding, only : c_ptr, c_char
+    type(c_ptr)                   :: client_to_string_c
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/fortran/client/client_interfaces.inc:159]: I think we tried that last time (on the get_last_error() PR) and it didn't work

File: tests/fortran/client_test_dataset.F90 [hunk 7]
```diff
@@ -195,6 +196,11 @@ program main
...
+  call dataset%print_dataset()
+  call dataset%print_dataset(stderr)
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply tests/fortran/client_test_dataset.F90:202]: Will do

[the remark this answers is not in the document]
Comment (billschereriii) [reply tests/fortran/client_test_dataset.F90:202]: Done

File: tests/fortran/client_test_dataset.F90 (hunk 5 above)
Comment (billschereriii) [reply tests/fortran/client_test_dataset.F90]: Done

**Review by MattToast — APPROVED** [review]
LGTM!!

File: src/cpp/dataset.cpp [hunk 8]
```diff
@@ -367,3 +370,41 @@ TensorBase* DataSet::_get_tensorbase_obj(const std::string& name)
...
 }
+
+// Create a string representation of the DataSet
+std::string DataSet::to_string() const
+{
+    std::string result;
+    result = "DataSet (" + _lname + "):\n";
+
+    // Tensors
+    result += "Tensors:\n";
```
Comment (mellis13) [inline src/cpp/dataset.cpp:381]: Very minor, but if number or tensors or number of metadata fields is 0, should we print something like ``None`` under the heading:

```
Metadata:
  None
```

So the message looks more complete when one is empty?

**Review by mellis13 — APPROVED** [review]
Just one minor question/suggestion.  Otherwise, great addition to the library and LGTM!

File: src/cpp/dataset.cpp (hunk 8 above)
Comment (billschereriii) [reply src/cpp/dataset.cpp:381]: That's a great suggestion!

**Pull request merged**